genesis/src/lockups.rs

```diff
@@ -0,0 +1,160 @@
+//! lockups generator
+use solana_sdk::{clock::Epoch, epoch_schedule::EpochSchedule, timing::years_as_slots};
+use std::time::Duration;
+
+#[derive(Debug, Default, Clone)]
+pub struct Lockups {
+    /// where in iteration progress I am
+    expiration: usize,
+    /// number of expirations post cliff
+    expirations: usize,
+    /// fractionage offered to f() last time
+    prev_fraction: f64,
+    /// portion of total available at cliff
+    cliff_fraction: f64,
+    /// time of cliff, in epochs
+    cliff_epoch: Epoch,
+    /// number of post-cliff events
+    expiration_fraction: f64,
+    /// time between each post-cliff event
+    expiration_epochs: Epoch,
+}
+
+impl Lockups {
+    pub fn by_years(
+        cliff_fraction: f64,   // first cliff fraction
+        cliff_year: f64,       // first cliff time, in years
+        expirations: usize,    // number of follow-on expirations
+        expiration_years: f64, // years between each following expiration
+        epoch_schedule: &EpochSchedule,
+        tick_duration: &Duration,
+        ticks_per_slot: u64,
+    ) -> Self {
+        let cliff_slot = years_as_slots(cliff_year, tick_duration, ticks_per_slot) as u64;
```

Comment:
I'm having a hard time reading this function. It looks like it's converting from durations to absolute heights, which is only possible if the duration started at zero (genesis). While that's a fine assumption for the genesis accounts, it seems unnecessarily restrictive.

Reply:
this lives in genesis, and it's the only consumer, and  it only needs zero-based epoch increments

MVP...

Reply:
I can't determine of the MVP is viable if I can't read the code.  `years_as_slots` is named such that it's only converting units ("as" implies a cast), but is actually converting a Duration (years) to a Date (slot_height) while also changing the units.

Reply:
How about this?

```rust
const GENESIS_SLOT_HEIGHT = 0;
...
let cliff_slots = years_as_slots(cliff_year, tick_duration, ticks_per_slot);
let cliff_slot_height = GENESIS_SLOT_HEIGHT + (cliff_slots as u64);
```

Reply:
is that really easier to read?

how about a comment "cliff_epoch is zero-based, these expirations all start from genesis, i.e. slot = 0"

Reply:
actually, we need unlocks to match Vesting dates, so I'm adding a "start year"

Reply:
scratch tha

Reply:
So `years_as_slots()` returns a duration, but `cliff_slot` implies `cliff_slot_height`. That conversion is almost implicit, except that you had to cast away from floating point numbers to get there.  Super confusing to read. The biggest problem is the name `cliff_slot`, which looks like a typo, given the function call (expected `cliff_slots`). But it's not a typo, it's that it was cast to height.

So yeah, the three line version is easier for me to read.  A commented, wrapper function `convert_years_from_genesis_to_slot_height(...) -> u64` would be even better.